### PR TITLE
Remove recipe ingredient and tile limits

### PR DESCRIPTION
**Pull request by Itorius**
Replaced recipe's required item and tile arrays with list => allowing for "infinite" number of ingredients/req. tiles.

There might be some uninteded behavior changes which need to be tested.

### CODE REVIEW

File: patches/tModLoader/Terraria/Recipe.Extensions.cs [hunk 1]
```diff
@@ -47,17 +37,7 @@ public partial class Recipe
...
-				if (groupID == -1) break;
-
-				if (groupID == id) {
-					return true;
-				}
-			}
-
-			return false;
-		}
+		public bool HasRecipeGroup(int id) => acceptedGroups.Any(group => group == id);
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Recipe.Extensions.cs:40]: `Contains`. Please

**Review by Chicken-Bones — COMMENTED** [review]
The code is good, but I honestly don't feel good about the maintenance cost here. Needs a second and third opinion.

**Comment by Mirsario** [comment]
There are 2 reasons to do this. The first one is obvious, mods like calamity and maybe fargo's love making giant recipes with every item in the game. The second one is that this will help people directly modify recipes, since I've heard that we're removing RecipeFinder & Editor. I agree on their removal.

File: patches/tModLoader/Terraria/Main.cs.patch [hunk 2]
```diff
@@ -2429,6 +2429,34 @@
...
+ 							UILinkPointNavigator.Shortcuts.CRAFT_CurrentIngridientsCount = num72 + 1;
+ 							break;
+@@ -30671,7 +_,8 @@
+ 						UILinkPointNavigator.Shortcuts.CRAFT_CurrentRecipeSmall = 1 + num72;
+ 						Microsoft.Xna.Framework.Color color5 = inventoryBack;
+ 						inventoryBack = new Microsoft.Xna.Framework.Color((byte)num75, (byte)num75, (byte)num75, (byte)num75);
+-						ItemSlot.Draw(spriteBatch, ref recipe[availableRecipe[focusRecipe]].requiredItem[num72], 22, new Vector2(num73, num74));
++						Item tempItem = recipe[availableRecipe[focusRecipe]].requiredItem[num72];
++						ItemSlot.Draw(spriteBatch, ref tempItem, 22, new Vector2(num73, num74));
+ 						inventoryBack = color5;
```
Comment (Mirsario) [inline patches/tModLoader/Terraria/Main.cs.patch:2448]: Shouldn't you set the list entry's value to tempitem here?

File: patches/tModLoader/Terraria/ModLoader/RecipeEditor.cs [hunk 3]
```diff
@@ -28,17 +28,17 @@ public class RecipeEditor
...
-					return;
-				}
-				if (recipe.requiredItem[k].type == itemID) {
-					recipe.requiredItem[k].stack += stack;
-					return;
-				}
-			}
+			// for (int k = 0; k < Recipe.maxRequirements; k++) {
+			// 	if (recipe.requiredItem[k].type == 0) {
+			// 		recipe.requiredItem[k].SetDefaults(itemID, false);
```
Comment (Mirsario) [inline patches/tModLoader/Terraria/ModLoader/RecipeEditor.cs:33]: Use comment blocks kthx

File: patches/tModLoader/Terraria/Recipe.cs.patch [hunk 4]
```diff
@@ -43,61 +53,102 @@
...
++			foreach (int num in acceptedGroups) {
 +				if (RecipeGroup.recipeGroups[num].ContainsItem(type)) {
  					theText = RecipeGroup.recipeGroups[num].GetText();
  					return true;
  				}
-@@ -82,20 +_,22 @@
+ 			}
+-
++			
+ 			theText = "";
```
Comment (Mirsario) [inline patches/tModLoader/Terraria/Recipe.cs.patch:100]: whitespace

**Review by Mirsario — CHANGES_REQUESTED** [review]

File: patches/tModLoader/Terraria/Main.cs.patch (hunk 2 above)
Comment (Itorius) [reply patches/tModLoader/Terraria/Main.cs.patch:2448]: this is just for drawing the crafting menu, the actual item in the slot won't change

File: patches/tModLoader/Terraria/ModLoader/RecipeEditor.cs (hunk 3 above)
Comment (Itorius) [reply patches/tModLoader/Terraria/ModLoader/RecipeEditor.cs:33]: RecipeEditor will get removed so I'm not worried about this

File: patches/tModLoader/Terraria/Recipe.Extensions.cs (hunk 1 above)
Comment (Itorius) [reply patches/tModLoader/Terraria/Recipe.Extensions.cs:40]: Woops, will fix in the main branch once this PR gets merged

Comment (Mirsario) [reply patches/tModLoader/Terraria/Recipe.Extensions.cs]: Just fixed this.

**Review by Mirsario — APPROVED** [review]

**Pull request merged**